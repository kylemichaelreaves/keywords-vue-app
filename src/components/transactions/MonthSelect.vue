<template>
    <el-select
        ref="selectComponent"
        v-model="selectedMonth"
        placeholder="select month"
        @update:model-value="$emit('update:selectedMonth', $event)"
        clearable
        filterable
    >
        <el-option
            v-for="option in transformedData"
            :key="option.value"
            :label="option.label"
            :value="option.value"
        />
    </el-select>
</template>

<script lang="ts">
import {computed, defineComponent, ref} from 'vue'
import {useMonths} from "../../api/hooks/transactions/useMonths";
import {ElOption, ElSelect} from "element-plus";
import {MonthYear} from "../../types";

export default defineComponent({
    name: "MonthSelect",
    components: {ElSelect, ElOption},
    emits: ['update:selectedMonth'],
    props: {
        selectedMonth: {
            type: String,
            default: ''
        }
    },
    setup(props) {

        const selectedMonth = props.selectedMonth

        const {data, isFetching, isLoading, isError, error} = useMonths()

        const transformedData = computed(() => {
            if (!data.value) {
                return []
            }
            return data.value.map((item: MonthYear) => ({
                value: item.month_year,
                label: item.month_year,
            }));
        });

        return {
            data,
            transformedData,
            isFetching,
            isLoading,
            isError,
            error,
            selectedMonth
        }
    }
})
</script>

<style scoped>

</style>
