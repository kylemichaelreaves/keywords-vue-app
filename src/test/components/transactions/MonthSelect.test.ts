import {vi, test} from 'vitest'
import {fireEvent, waitFor, screen, render} from "@testing-library/vue";
import MonthSelect from "../../../components/transactions/MonthSelect.vue";
import {renderWithQueryClient} from "../../renderWithQueryClient";
import {ElOption, ElSelect} from "element-plus";
import {VueQueryPlugin} from "@tanstack/vue-query";
import {mount} from "@vue/test-utils";
import {nextTick} from "vue";


describe('MonthsSelect', () => {


    test('should render', () => {
        const wrapper = mount(MonthSelect, {
            global: {
                plugins: [ElSelect, ElOption, VueQueryPlugin],
            }
        })

        expect(wrapper.exists()).toBe(true)
    })

    test('should render the component and its options', async () => {

        render(MonthSelect, {
            global: {
                plugins: [ElSelect, VueQueryPlugin],
            }
        })


        await waitFor(() => {
            expect(screen.getByPlaceholderText('select month')).toBeInTheDocument()
            expect(screen.getByText('11/2022')).toBeInTheDocument()
            expect(screen.getByText('12/2022')).toBeInTheDocument()
            expect(screen.getByText('01/2023')).toBeInTheDocument()
        })
    })

    test('clearable should be true', async () => {
        const wrapper = mount(MonthSelect, {
            global: {
                plugins: [ElSelect, VueQueryPlugin],
            }
        })

        expect(wrapper.vm.$refs.selectComponent.clearable).toBe(true)
    })

    test('should clear the selected value when the clearable button is clicked', async () => {
        const wrapper = mount(MonthSelect, {
            global: {
                plugins: [ElSelect, VueQueryPlugin],
            }
        })

        await wrapper.findComponent(ElSelect).setValue('11/2022')

        await nextTick()

        await wrapper.findComponent(ElSelect).find('.el-select__caret').trigger('click')

        await nextTick()

        await wrapper.findComponent(ElSelect).find('.el-select-dropdown__item').trigger('click')

        await nextTick()

        expect(wrapper.vm.$refs.selectComponent.modelValue).toBe('')

        // const select = screen.findAllByDisplayValue(/select month/i)

        // await fireEvent.update(select, '11/2022')

        // await waitFor(() => {
        //     expect(screen.getByText('11/2022')).toBeInTheDocument()
        // })

        // screen.getByText('11/2022').click()


    });
})
