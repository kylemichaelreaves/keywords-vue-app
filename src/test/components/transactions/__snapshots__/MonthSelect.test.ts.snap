// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MonthsSelect > should render 1`] = `""`;
